# PROJECT ASSIGNMENT #2
### GROUP MEMBERS: Hannah Seow, Sophia Porter, Neil Chen, Will Wang

In [20]:
import pandas as pd
import numpy as np
df_indicators = pd.read_csv('country_indicators.csv')
df_indicators

Unnamed: 0  sowc_demographics__population-thousands-2021_total  \
0             0                                         40099.4620    
1             1                                          2854.7100    
2             2                                         44177.9685    
3             3                                            79.0340    
4             4                                         34503.7735    
..          ...                                                ...    
213         213                                                NaN    
214         214                                                NaN    
215         215                                                NaN    
216         216                                                NaN    
217         217                                                NaN    

     sowc_demographics__population-thousands-2021_under-18  \
0                                           20297.8860       
1                                             574.3875       
2                                           15526.2050       
3                                              12.7765       
4                                           17832.9760       
..                                                 ...       
213                                                NaN       
214                                                NaN       
215                                                NaN       
216                                                NaN       
217                                                NaN       

     sowc_demographics__population-thousands-2021_under-5  \
0                                            6490.5540      
1                                             141.3930      
2                                            4870.8425      
3                                               2.5385      
4                                            5983.4065      
..                                                 ...      
213                                                NaN      
214                                                NaN      
215                                                NaN      
216                                                NaN      
217                                                NaN      

     sowc_demographics__annual-population-growth-rate_2000-2020  \
0                                             3.286824            
1                                             0.496681            
2                                             1.642661            
3                                             0.770116            
4                                             3.392806            
..                                                 ...            
213                                                NaN            
214                                                NaN            
215                                                NaN            
216                                                NaN            
217                                                NaN            

     sowc_demographics__annual-population-growth-rate_2020-2030-a  \
0                                             2.325169              
1                                             0.248324              
2                                             1.237367              
3                                             0.437249              
4                                             2.684447              
..                                                 ...              
213                                                NaN              
214                                                NaN              
215                                                NaN              
216                                                NaN              
217                                                NaN              

     sowc_demographics__annual-number-of

In [21]:
df_preds = pd.read_csv('test_predictions.csv')
df_preds

yearmonth fips  y_pred_transformer  y_pred_proba_transformer  \
0       202211   FJ               False                  0.183897   
1       202212   FJ               False                  0.267831   
2       202211   TZ               False                  0.482585   
3       202212   TZ               False                  0.187792   
4       202301   TZ                True                  0.539319   
..         ...  ...                 ...                       ...   
359     202211   MJ               False                  0.182196   
360     202212   MJ               False                  0.203236   
361     202211   TD                True                  0.527107   
362     202212   TD                True                  0.555677   
363     202301   TD                True                  0.565700   

     y_true_transformer  y_pred_xgboost  y_pred_proba_xgboost  y_true_xgboost  \
0                 False           False              0.066500           False   
1                 False           False              0.099643           False   
2                 False            True              0.704086            True   
3                 False            True              0.638444            True   
4                  True            True              0.608380           False   
..                  ...             ...                   ...             ...   
359               False           False              0.079453           False   
360               False           False              0.060189           False   
361               False            True              0.697625            True   
362               False            True              0.729246           False   
363                True            True              0.591722           False   

     y_pred_ffnn  y_pred_proba_ffnn  y_true_ffnn iso3  
0          False           0.409958        False  FJI  
1          False           0.406696        False  FJI  
2           True           0.545236        False  TZA  
3           True           0.534560        False  TZA  
4           True           0.538583         True  TZA  
..           ...                ...          ...  ...  
359        False           0.291874        False  MNE  
360        False           0.300321        False  MNE  
361        False           0.335496        False  TTO  
362        False           0.324000        False  TTO  
363        False           0.332455         True  TTO  

[364 rows x 12 columns]

In [22]:
df = df_preds.merge(df_indicators, left_on='iso3', right_on='iso3', how='inner')
df

yearmonth fips  y_pred_transformer  y_pred_proba_transformer  \
0       202211   FJ               False                  0.183897   
1       202212   FJ               False                  0.267831   
2       202211   TZ               False                  0.482585   
3       202212   TZ               False                  0.187792   
4       202301   TZ                True                  0.539319   
..         ...  ...                 ...                       ...   
359     202211   MJ               False                  0.182196   
360     202212   MJ               False                  0.203236   
361     202211   TD                True                  0.527107   
362     202212   TD                True                  0.555677   
363     202301   TD                True                  0.565700   

     y_true_transformer  y_pred_xgboost  y_pred_proba_xgboost  y_true_xgboost  \
0                 False           False              0.066500           False   
1                 False           False              0.099643           False   
2                 False            True              0.704086            True   
3                 False            True              0.638444            True   
4                  True            True              0.608380           False   
..                  ...             ...                   ...             ...   
359               False           False              0.079453           False   
360               False           False              0.060189           False   
361               False            True              0.697625            True   
362               False            True              0.729246           False   
363                True            True              0.591722           False   

     y_pred_ffnn  y_pred_proba_ffnn  ...  \
0          False           0.409958  ...   
1          False           0.406696  ...   
2           True           0.545236  ...   
3           True           0.534560  ...   
4           True           0.538583  ...   
..           ...                ...  ...   
359        False           0.291874  ...   
360        False           0.300321  ...   
361        False           0.335496  ...   
362        False           0.324000  ...   
363        False           0.332455  ...   

     fsi_e3:_human_flight_and_brain_drain fsi_e2:_economic_inequality  \
0                                     7.6                         5.0   
1                                     7.6                         5.0   
2                                     6.4                         7.3   
3                                     6.4                         7.3   
4                                     6.4                         7.3   
..                                    ...                         ...   
359                                   5.4                         2.4   
360                                   5.4                         2.4   
361                                   6.8                         4.7   
362                                   6.8                         4.7   
363                                   6.8                         4.7   

     fsi_e1:_economy  fsi_p1:_state_legitimacy  fsi_p2:_public_services  \
0                7.5                       6.1                      4.2   
1                7.5                       6.1                      4.2   
2                6.2                       6.9                      8.4   
3                6.2                       6.9                      8.4   
4                6.2                       6.9                      8.4   
..               ...                       ...                      ...   
359              5.8                       3.9                      3.8   
360              5.8                       3.9                      3.8   
361              3.9                       3.5                      4.3   
362              3.9                       3.5 

In [23]:
from sklearn import metrics
import seaborn as sns

column_defining_groups = 'fsi_category'

chosen_metric = 'accuracy_score'

metric_function = getattr(metrics, chosen_metric)

chosen_metric_for_chosen_groups = list()
chosen_metric_back_into_original_data = {'ffnn': pd.Series(index=df.index, dtype=np.float64), 
                                         'xgboost': pd.Series(index=df.index, dtype=np.float64), 
                                         'transformer': pd.Series(index=df.index, dtype=np.float64)}
    
for g, rows in df.groupby(column_defining_groups):
    for model in ['ffnn', 'xgboost', 'transformer']:        
        chosen_metric_value = metric_function(rows[f"y_true_{model}"], rows[f"y_pred_{model}"]) 
        
        chosen_metric_back_into_original_data[model][rows.index] = chosen_metric_value
        chosen_metric_for_chosen_groups.append((model, g, chosen_metric_value))

cm_bigger_better = sns.light_palette("teal", as_cmap=True)

styler = ( pd.DataFrame(chosen_metric_for_chosen_groups, 
                        columns=('model', column_defining_groups, chosen_metric))
            .sort_values(column_defining_groups).style
            .background_gradient(cmap=cm_bigger_better)
            .format(precision=5).hide(axis="index"))
styler

In [54]:
import plotly.express as px

fig = px.choropleth(df, 
                    locations='iso3',
                    color='fsi_category',
                    hover_name='iso3',
                    color_continuous_scale=px.colors.sequential.Plasma)
fig.show()

# ***HYPOTHESIS TEST*** $1$
### (between all three models)


We believe that the probability that all three models should accurately predict a true positive (correctly predicted escalation) or a true negative (correctly predicted no escalation) for the fsi category 'sustainable' is 0.5. This is because the model will either have to choose if a country has conflict escalation (positive) or no escalation (negative) and therefore, has a 0.5 chance that it is true and a 0.5 chance it is false. 

- Let t be 'correctly predicted escalation and no escalation' under **transformer** model for fsi category 'Sustainable'
- Let f be 'correctly predicted escalation and no escalation' under **ffnn** model for fsi category 'Sustainable' 
- Let x be 'correctly predicted escalation and no escalation' under **xgboost** model for fsi category 'Sustainable'
- Let c be 'correctly predicted escalation and no escalation' under **all three models** for fsi category 'Sustainable'

$H_0: p_c = p_t + p_f + p_x = 0.5$

$H_1$: Not $H_0$

In [25]:
df.fsi_category.value_counts() #to count the number of countries with sustainable fsi_category

Warning        171
Stable          88
Alert           70
Sustainable     35
Name: fsi_category, dtype: int64

## To Calculate ($p_t$): The Proportion of 'Correctly Predicted Escalation and No Escalation under transformer model for fsi category 'Sustainable' 

In [26]:
threshold = 0.5 # using default classification threshold

df_preds['transformer_classification_performance_outcome'] = 0.5

tmpt = df_preds['transformer_classification_performance_outcome'].copy()
TP_pos_pred_correct = df_preds.y_true_transformer & (df_preds.y_pred_proba_transformer>threshold)
tmpt[TP_pos_pred_correct] = "correctly predicted escalation"
TN_neg_pred_correct = (~df_preds.y_true_transformer) & (df_preds.y_pred_proba_transformer<=threshold)
tmpt[TN_neg_pred_correct] = "correctly predicted no escalation"

df_preds['transformer_classifcation_performance_outcome'] = tmpt
df_preds[['y_true_transformer','y_pred_transformer','transformer_classification_performance_outcome']]

y_true_transformer  y_pred_transformer  \
0                 False               False   
1                 False               False   
2                 False               False   
3                 False               False   
4                  True                True   
..                  ...                 ...   
359               False               False   
360               False               False   
361               False                True   
362               False                True   
363                True                True   

     transformer_classification_performance_outcome  
0                                               0.5  
1                                               0.5  
2                                               0.5  
3                                               0.5  
4                                               0.5  
..                                              ...  
359                                             0.5  
360                                             0.5  
361                                             0.5  
362                                             0.5  
363                                             0.5  

[364 rows x 3 columns]

In [27]:
pt_TP_pos_pred_correct = (tmpt[TP_pos_pred_correct].groupby(df.fsi_category))
pt_TP_pos_pred_correct.value_counts()/364

fsi_category  transformer_classification_performance_outcome
Alert         correctly predicted escalation                    0.024725
Stable        correctly predicted escalation                    0.008242
Sustainable   correctly predicted escalation                    0.002747
Warning       correctly predicted escalation                    0.052198
Name: transformer_classification_performance_outcome, dtype: float64

In [28]:
pt_TN_neg_pred_correct = (tmpt[TN_neg_pred_correct].groupby(df.fsi_category))
pt_TN_neg_pred_correct.value_counts()/364

fsi_category  transformer_classification_performance_outcome
Alert         correctly predicted no escalation                 0.065934
Stable        correctly predicted no escalation                 0.164835
Sustainable   correctly predicted no escalation                 0.087912
Warning       correctly predicted no escalation                 0.211538
Name: transformer_classification_performance_outcome, dtype: float64

### $p_t = 0.002747 + 0.087912 = 0.090659$

Let $p_t$ be the observed proportion of true positives under the ffnn model for sustainable fsi category plus the proportion of true negatives under the transformer model for sustainable fsi category

## To Calculate ($p_f$): The Proportion of 'Correctly Predicted Escalation and No Escalation under ffnn model for fsi category 'Sustainable' 

In [29]:
threshold = 0.5 # using default classification threshold

df_preds['ffnn_classification_performance_outcome'] = 0.5

tmpf = df_preds['ffnn_classification_performance_outcome'].copy()
TP_pos_pred_correct = df_preds.y_true_ffnn & (df_preds.y_pred_proba_ffnn>threshold)
tmpf[TP_pos_pred_correct] = "correctly predicted escalation"
TN_neg_pred_correct = (~df_preds.y_true_ffnn) & (df_preds.y_pred_proba_ffnn<=threshold)
tmpf[TN_neg_pred_correct] = "correctly predicted no escalation"

df_preds['ffnn_classifcation_performance_outcome'] = tmpf
df_preds[['y_true_ffnn','y_pred_ffnn','ffnn_classification_performance_outcome']]

y_true_ffnn  y_pred_ffnn  ffnn_classification_performance_outcome
0          False        False                                      0.5
1          False        False                                      0.5
2          False         True                                      0.5
3          False         True                                      0.5
4           True         True                                      0.5
..           ...          ...                                      ...
359        False        False                                      0.5
360        False        False                                      0.5
361        False        False                                      0.5
362        False        False                                      0.5
363         True        False                                      0.5

[364 rows x 3 columns]

In [30]:
pf_TP_pos_pred_correct = (tmpf[TP_pos_pred_correct].groupby(df.fsi_category))
pf_TP_pos_pred_correct.value_counts()/364

fsi_category  ffnn_classification_performance_outcome
Alert         correctly predicted escalation             0.021978
Warning       correctly predicted escalation             0.041209
Name: ffnn_classification_performance_outcome, dtype: float64

In [31]:
pf_TN_neg_pred_correct = (tmpf[TN_neg_pred_correct].groupby(df.fsi_category))
pf_TN_neg_pred_correct.value_counts()/364

fsi_category  ffnn_classification_performance_outcome
Alert         correctly predicted no escalation          0.052198
Stable        correctly predicted no escalation          0.217033
Sustainable   correctly predicted no escalation          0.090659
Warning       correctly predicted no escalation          0.304945
Name: ffnn_classification_performance_outcome, dtype: float64

### $p_f = 0 + 0.090659 = 0.090659$

Let $p_f$ be the observed proportion of true positives under the ffnn model for sustainable fsi category plus the proportion of true negatives under the ffnn model for sustainable fsi category

## To Calculate ($p_x$): The Proportion of 'Correctly Predicted Escalation and No Escalation under xgboost model for fsi category 'Sustainable' 

In [32]:
threshold = 0.5 # using default classification threshold

df_preds['xgboost_classification_performance_outcome'] = 0.5

tmpx = df_preds['xgboost_classification_performance_outcome'].copy()
TP_pos_pred_correct = df_preds.y_true_xgboost & (df_preds.y_pred_proba_xgboost>threshold)
tmpx[TP_pos_pred_correct] = "correctly predicted escalation"
TN_neg_pred_correct = (~df_preds.y_true_xgboost) & (df_preds.y_pred_proba_xgboost<=threshold)
tmpx[TN_neg_pred_correct] = "correctly predicted no escalation"

df_preds['xgboost_classifcation_performance_outcome'] = tmpx
df_preds[['y_true_xgboost','y_pred_xgboost','xgboost_classification_performance_outcome']]

y_true_xgboost  y_pred_xgboost  \
0             False           False   
1             False           False   
2              True            True   
3              True            True   
4             False            True   
..              ...             ...   
359           False           False   
360           False           False   
361            True            True   
362           False            True   
363           False            True   

     xgboost_classification_performance_outcome  
0                                           0.5  
1                                           0.5  
2                                           0.5  
3                                           0.5  
4                                           0.5  
..                                          ...  
359                                         0.5  
360                                         0.5  
361                                         0.5  
362                                         0.5  
363                                         0.5  

[364 rows x 3 columns]

In [33]:
px_TP_pos_pred_correct = (tmpx[TP_pos_pred_correct].groupby(df.fsi_category))
px_TP_pos_pred_correct.value_counts()/364

fsi_category  xgboost_classification_performance_outcome
Alert         correctly predicted escalation                0.024725
Stable        correctly predicted escalation                0.010989
Warning       correctly predicted escalation                0.068681
Name: xgboost_classification_performance_outcome, dtype: float64

In [34]:
px_TN_neg_pred_correct = (tmpx[TN_neg_pred_correct].groupby(df.fsi_category))
px_TN_neg_pred_correct.value_counts()/364

fsi_category  xgboost_classification_performance_outcome
Alert         correctly predicted no escalation             0.032967
Stable        correctly predicted no escalation             0.170330
Sustainable   correctly predicted no escalation             0.079670
Warning       correctly predicted no escalation             0.118132
Name: xgboost_classification_performance_outcome, dtype: float64

### $p_x = 0 + 0.079670 = 0.079670$

Let $p_x$ be the observed proportion of true positives under the xgboost model for sustainable fsi category plus the proportion of true negatives under the xgboost model for sustainable fsi category

### Therefore, the observed test statistic would be the total observed proportion of true positives and true negatives for the sustainable fsi_category for all three models which is given by the equation:

$p_c$ (observed) = $p_x$ (observed) + $p_f$ (observed) + $p_t$ (observed)

$p_c$ (observed) = 0.079670 + 0.090659 + 0.090659

$p_c$ (observed) = 0.260988

In [35]:
simulated_samples, n = 10000, 35 #n=35 since there are 35 countries with the sustainable fsi_category (as seen in ln6)
p_hats = [] 
p = 0.5 # Null Hypothesis assumed proportion
np.random.seed(100) 
for i in range(simulated_samples): 
    sample = np.random.choice(['True Positive and True Negative', 'False Positive and False Negative'], size=n, replace=True, p=[p, 1-p])
    sample_prop = (sample == 'True Positive and True Negative').sum() / n
    p_hats.append(sample_prop)

observed_test_stat = 0.260988
p = 0.5
num_more_extreme = (abs(np.array(p_hats) - p) >= abs(observed_test_stat - p)).sum()
p_value = num_more_extreme / simulated_samples
p_value

0.0051

Therefore, since we got a **p-value of 0.0051**, this shows that there is **very strong evidence against the null** which means that the probability that all three models (xgboost, ffnn and transformer) accurately predicted escalation or no escalation is **not equal to 0.5**. In reality, it can be seen that the probability of true positives and true negatives for the fsi category 'sustainable' is approximately 0.905 (the average of the accuracy score for all three models for the given category).

# This concludes the Hypothesis Test $1$.

In [55]:
#Visual aid for 'wbi_income_group' catagory for acuracy

import numpy as np
from sklearn import metrics
import seaborn as sns

column_defining_groups = 'wbi_income_group'
chosen_metric = 'accuracy_score'

metric_function = getattr(metrics, chosen_metric)

chosen_metric_for_chosen_groups = list()
chosen_metric_back_into_original_data = {'ffnn': pd.Series(index=df.index, dtype=np.float64), 
                                         'xgboost': pd.Series(index=df.index, dtype=np.float64), 
                                         'transformer': pd.Series(index=df.index, dtype=np.float64)}
    
for g, rows in df.groupby(column_defining_groups):
    for model in ['ffnn', 'xgboost', 'transformer']:        
        chosen_metric_value = metric_function(rows[f"y_true_{model}"], rows[f"y_pred_{model}"])

        
        chosen_metric_back_into_original_data[model][rows.index] = chosen_metric_value
        chosen_metric_for_chosen_groups.append((model, g, chosen_metric_value))

cm_bigger_better = sns.light_palette("pink", as_cmap=True)
styler = ( pd.DataFrame(chosen_metric_for_chosen_groups, 
                        columns=('model', column_defining_groups, chosen_metric))
            .sort_values(column_defining_groups).style
            .background_gradient(cmap=cm_bigger_better)
            .format(precision=3).hide(axis="index"))
styler

In [56]:
#map of wbi_income_group desplaying rows in colours
import plotly.express as px
fig = px.choropleth(df,
                    locations='iso3',
                    color='wbi_income_group',
                    hover_name='iso3',
                    color_continuous_scale=px.colors.sequential.Plasma)
fig.show()

In [ ]:
#confidence intervals
import numpy as np
from sklearn import metrics
import seaborn as sns
import pandas as pd

df_visualization = pd.DataFrame(chosen_metric_for_chosen_groups, columns=('model', column_defining_groups, chosen_metric))

def calculate_confidence_interval(data):
    lower_bound = np.quantile(data, 0.025)
    upper_bound = np.quantile(data, 0.975)
    return lower_bound, upper_bound
conf_intervals = df_visualization.groupby(column_defining_groups)[chosen_metric].apply(calculate_confidence_interval)
conf_intervals

## Confidance interval of low income
 ***(0.2620689655172414, 0.5405172413793103)***

Roughly: ***(0.26, 0.54)***

# ***HYPOTHESIS TEST*** $2$
The hypothesis assumes each model's (transformer, ffnn, and xgboost) error score of low income in countries is the same.

***Introducing valuables:***

- Let t be 'wrongly predicted escalation' under **transformer** model for fsi category 'Low income'
- Let f be 'wrongly predicted escalation' under **ffnn** model for fsi category 'Low income' 
- Let x be 'wrongly predicted escalation' under **xgboost** model for fsi category 'Low income'
- Let c be 'wrongly predicted escalation ' under **all three models** for fsi category 'Low income'

# $H_0: p_c = p_t + p_f + p_x = 0.5$

# Alternative Hypothesis
the alternative hypothesis assumes that at least on the the models has a different score

# $H_A: H_1: H_0 {is false}$

In [37]:
#this code finds all the catagories of the df that have dtype 'object'
list(df_indicators.select_dtypes(['object']).columns)

['iso3',
 'hdr_hdicode',
 'hdr_region',
 'wbi_income_group',
 'wbi_lending_category',
 'wbi_other_(emu_or_hipc)',
 'fsi_rank',
 'fsi_category']

In [38]:
#this code counts the number of rows that have data for low income
df_indicators.wbi_income_group.value_counts()

High income            73
Upper middle income    53
Lower middle income    47
Low income             24
Name: wbi_income_group, dtype: int64

In [41]:
# Prediction Classification 'falsly predicted'(transformer model)

threshold = 0.5 # using default classification threshold


df_preds['transformer_classifcation_performance_outcome'] = 0.5

tmpt = df_preds['transformer_classifcation_performance_outcome'].copy()

FP_pos_pred_wrong = (~df_preds.y_true_transformer) & (df_preds.y_pred_proba_transformer>threshold)
tmpt[FP_pos_pred_wrong] = "wrongly predicted escalation"
FN_neg_pred_wrong = df_preds.y_true_transformer & (df_preds.y_pred_proba_transformer<=threshold)
tmpt[FN_neg_pred_wrong] = "wrongly predicted no escalation"

df_preds['transformer_classifcation_performance_outcome'] = tmpt
df_preds[['y_true_transformer','y_pred_transformer','transformer_classifcation_performance_outcome']]

y_true_transformer  y_pred_transformer  \
0                 False               False   
1                 False               False   
2                 False               False   
3                 False               False   
4                  True                True   
..                  ...                 ...   
359               False               False   
360               False               False   
361               False                True   
362               False                True   
363                True                True   

    transformer_classifcation_performance_outcome  
0                                             0.5  
1                                             0.5  
2                                             0.5  
3                                             0.5  
4                                             0.5  
..                                            ...  
359                                           0.5  
360                                           0.5  
361                  wrongly predicted escalation  
362                  wrongly predicted escalation  
363                                           0.5  

[364 rows x 3 columns]

In [42]:
#wrongly predicted positives (false positives)
wrong_predictions_t = (tmpt[FP_pos_pred_wrong].groupby(df.wbi_income_group))
wrong_predictions_t.value_counts()/364

wbi_income_group     transformer_classifcation_performance_outcome
High income          wrongly predicted escalation                     0.041209
Low income           wrongly predicted escalation                     0.063187
Lower middle income  wrongly predicted escalation                     0.107143
Upper middle income  wrongly predicted escalation                     0.109890
Name: transformer_classifcation_performance_outcome, dtype: float64

In [43]:
#wrongly predicted negatives (false negatives)
wrong_predictions_t = (tmpt[FN_neg_pred_wrong].groupby(df.wbi_income_group))
wrong_predictions_t.value_counts()/364

wbi_income_group     transformer_classifcation_performance_outcome
High income          wrongly predicted no escalation                  0.013736
Low income           wrongly predicted no escalation                  0.008242
Lower middle income  wrongly predicted no escalation                  0.016484
Upper middle income  wrongly predicted no escalation                  0.021978
Name: transformer_classifcation_performance_outcome, dtype: float64

 #### Wrongly predicting (false positive and false negative) ###
 '***low income***' in countries with ***transformer*** model w false positive= ***0.063187*** and false negative is= ***0.008242***
 
 $p_t= 0.063187+0.008242$ = ***0.145607*** (observed proportions)

In [44]:
# Prediction Classification 'falsly predicted'(ffnn model)

threshold = 0.5 
df_preds['ffnn_classifcation_performance_outcome'] = 0.5

tmp = df_preds['ffnn_classifcation_performance_outcome'].copy()

FP_pos_pred_wrong = (~df_preds.y_true_ffnn) & (df_preds.y_pred_proba_ffnn>threshold)
tmp[FP_pos_pred_wrong] = "wrongly predicted escalation"
FN_neg_pred_wrong = df_preds.y_true_ffnn & (df_preds.y_pred_proba_ffnn>threshold)
tmp[FN_neg_pred_wrong] = "wrongly predicted no escalation"

df_preds['ffnn_classifcation_performance_outcome'] = tmp
df_preds[['y_true_ffnn','y_pred_proba_ffnn','ffnn_classifcation_performance_outcome']]

y_true_ffnn  y_pred_proba_ffnn ffnn_classifcation_performance_outcome
0          False           0.409958                                    0.5
1          False           0.406696                                    0.5
2          False           0.545236           wrongly predicted escalation
3          False           0.534560           wrongly predicted escalation
4           True           0.538583        wrongly predicted no escalation
..           ...                ...                                    ...
359        False           0.291874                                    0.5
360        False           0.300321                                    0.5
361        False           0.335496                                    0.5
362        False           0.324000                                    0.5
363         True           0.332455                                    0.5

[364 rows x 3 columns]

In [45]:
#wrongly predicted positives (false positives)
wrong_predictions_f = (tmp[FP_pos_pred_wrong].groupby(df.wbi_income_group))
wrong_predictions_f.value_counts()/364

wbi_income_group     ffnn_classifcation_performance_outcome
Low income           wrongly predicted escalation              0.118132
Lower middle income  wrongly predicted escalation              0.065934
Name: ffnn_classifcation_performance_outcome, dtype: float64

In [46]:
#wrongly predicted negatives (false negatives)
wrong_predictions_f = (tmp[FN_neg_pred_wrong].groupby(df.wbi_income_group))
wrong_predictions_f.value_counts()/364

wbi_income_group     ffnn_classifcation_performance_outcome
Low income           wrongly predicted no escalation           0.038462
Lower middle income  wrongly predicted no escalation           0.024725
Name: ffnn_classifcation_performance_outcome, dtype: float64

In [47]:
# Prediction Classification 'falsly predicted'(xgboost model)

threshold = 0.5 # using default classification threshold

# using transformer for this example demonstration, as opposed to
# df_preds.y_true_ffnn and df_preds.y_pred_proba_ffnn>threshold
# df_preds.y_true_xgboost and df_preds.y_pred_proba_xgboost>threshold
df_preds['xgboost_classifcation_performance_outcome'] = 0.5

tmpr = df_preds['xgboost_classifcation_performance_outcome'].copy()

FP_pos_pred_wrong = (~df_preds.y_true_xgboost) & (df_preds.y_pred_proba_xgboost>threshold)
tmpr[FP_pos_pred_wrong] = "wrongly predicted escalation"
FN_neg_pred_wrong = df_preds.y_true_xgboost & (df_preds.y_pred_proba_xgboost<=threshold)
tmpr[FN_neg_pred_wrong] = "wrongly predicted no escalation"

df_preds['xgboost_classifcation_performance_outcome'] = tmp
df_preds[['y_true_xgboost','y_pred_proba_xgboost','xgboost_classifcation_performance_outcome']]

y_true_xgboost  y_pred_proba_xgboost  \
0             False              0.066500   
1             False              0.099643   
2              True              0.704086   
3              True              0.638444   
4             False              0.608380   
..              ...                   ...   
359           False              0.079453   
360           False              0.060189   
361            True              0.697625   
362           False              0.729246   
363           False              0.591722   

    xgboost_classifcation_performance_outcome  
0                                         0.5  
1                                         0.5  
2                wrongly predicted escalation  
3                wrongly predicted escalation  
4             wrongly predicted no escalation  
..                                        ...  
359                                       0.5  
360                                       0.5  
361                                       0.5  
362                                       0.5  
363                                       0.5  

[364 rows x 3 columns]

 #### Wrongly predicting (false positive and false negative) ###
 '***low income***' in countries with ***ffnn*** model with false positive= ***0.118132*** and false negative is= ***0.038462***
 
 $p_f= 0.118132+0.038462$ = ***0.156594*** (observed proportions)

In [48]:
# Prediction Classification "Correctness" (xgboost model)

threshold = 0.5 # using default classification threshold

# using transformer for this example demonstration, as opposed to
# df_preds.y_true_ffnn and df_preds.y_pred_proba_ffnn>threshold
# df_preds.y_true_xgboost and df_preds.y_pred_proba_xgboost>threshold
df_preds['xgboost_classifcation_performance_outcome'] = 0.5

tmpr = df_preds['xgboost_classifcation_performance_outcome'].copy()

FP_pos_pred_wrong = (~df_preds.y_true_xgboost) & (df_preds.y_pred_proba_xgboost>threshold)
tmpr[FP_pos_pred_wrong] = "wrongly predicted escalation"
FN_neg_pred_wrong = df_preds.y_true_xgboost & (df_preds.y_pred_proba_xgboost<=threshold)
tmpr[FN_neg_pred_wrong] = "wrongly predicted no escalation"

df_preds['xgboost_classifcation_performance_outcome'] = tmp
df_preds[['y_true_xgboost','y_pred_proba_xgboost','xgboost_classifcation_performance_outcome']]

y_true_xgboost  y_pred_proba_xgboost  \
0             False              0.066500   
1             False              0.099643   
2              True              0.704086   
3              True              0.638444   
4             False              0.608380   
..              ...                   ...   
359           False              0.079453   
360           False              0.060189   
361            True              0.697625   
362           False              0.729246   
363           False              0.591722   

    xgboost_classifcation_performance_outcome  
0                                         0.5  
1                                         0.5  
2                wrongly predicted escalation  
3                wrongly predicted escalation  
4             wrongly predicted no escalation  
..                                        ...  
359                                       0.5  
360                                       0.5  
361                                       0.5  
362                                       0.5  
363                                       0.5  

[364 rows x 3 columns]

In [49]:
#wrongly predicted positives (false positives)
wrong_predictions_f = (tmpr[FP_pos_pred_wrong].groupby(df.wbi_income_group))
wrong_predictions_f.value_counts()/364

wbi_income_group     xgboost_classifcation_performance_outcome
High income          wrongly predicted escalation                 0.041209
Low income           wrongly predicted escalation                 0.107143
Lower middle income  wrongly predicted escalation                 0.156593
Upper middle income  wrongly predicted escalation                 0.145604
Name: xgboost_classifcation_performance_outcome, dtype: float64

In [50]:
#wrongly predicted negatives (false negatives)
wrong_predictions_f = (tmpr[FN_neg_pred_wrong].groupby(df.wbi_income_group))
wrong_predictions_f.value_counts()/364

wbi_income_group     xgboost_classifcation_performance_outcome
High income          wrongly predicted no escalation              0.013736
Lower middle income  wrongly predicted no escalation              0.005495
Upper middle income  wrongly predicted no escalation              0.021978
Name: xgboost_classifcation_performance_outcome, dtype: float64

 #### Wrongly predicting (false positive and false negative) ###
 '***low income***' in countries with ***ffnn*** model with false positive= ***0.107143*** and false negative is= ***0***
 
 $p_x = 0.107143+0$ = ***0.107143*** (observed proportions)

## Total  of observed test statistic:

$p_c$ (observed) = $p_x$ (observed) + $p_f$ (observed) + $p_t$ (observed)

$p_c$ (observed) = 0.145607 + 0.156594 + 0.107143

$p_c$ (observed) = ***0.409344***

# Hypothesis test


In [51]:
simulated_samples, n = 10000, 35 #n=35 since there are 35 countries with the sustainable fsi_category (as seen in ln6)
p_hats = [] 
p = 0.5 # Null Hypothesis assumed proportion
np.random.seed(100) 
for i in range(simulated_samples): 
    sample = np.random.choice(['True Positive and True Negative', 'False Positive and False Negative'], size=n, replace=True, p=[p, 1-p])
    sample_prop = (sample == 'False Positive and False Negative').sum() / n
    p_hats.append(sample_prop)

observed_test_stat = 0.409344
p = 0.5
num_more_extreme = (abs(np.array(p_hats) - p) >= abs(observed_test_stat - p)).sum()
p_value = num_more_extreme / simulated_samples
p_value

0.3139

The hypothesis test concludes there is ***no evidance against*** the null hypothesis because the ***p-value*** is ***0.3139*** > 0.10. Concluding that there was no evidance to contridict the assumption that the models produce the same error results of 0.5.

# This Concludes Hypothesis Test $2$.